firewood/src/merkle/stream.rs

```diff
@@ -0,0 +1,505 @@
+                    visit_last,
+                    next_result,
+                    parents,
+                };
+
+                self.poll_next(_cx)
+            }
+
+            IteratorState::StartAtKey(key) => {
+                let root_node = merkle
+                    .get_node(*merkle_root)
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                let (next_result, parents) = {
+                    let (found_node, parents) = {
+                        let mut parents = vec![];
+
+                        let found_node = merkle
+                            .get_node_by_key_with_callbacks(
+                                root_node,
+                                &key,
+                                |addr, nib| {
+                                    parents.push((addr, nib));
+                                },
+                                |_, _| {},
+                            )
+                            .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                        (found_node, parents)
+                    };
+
+                    let mut parents = parents
+                        .into_iter()
+                        .map(|(addr, nib)| merkle.get_node(addr).map(|node| (node, nib)))
+                        .collect::<Result<Vec<_>, _>>()
+                        .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                    // means the keys definitely match
+                    if let Some(found_node) = found_node {
+                        let (value, visit_last) = match found_node.inner() {
+                            NodeType::Extension(_) => unreachable!(),
+
+                            NodeType::Branch(branch) => {
+                                let value = branch.value.as_ref().map(|v| v.to_vec());
+                                let visit_last = true;
+                                parents.push((found_node, 0));
+
+                                (value, visit_last)
+                            }
+
+                            NodeType::Leaf(leaf) => {
+                                let value = leaf.data.to_vec();
+                                let visit_last = false;
+
+                                (Some(value), visit_last)
+                            }
+                        };
+
+                        let next_result = value.map(|value| (std::mem::take(key), value));
+
+                        self.key_state = IteratorState::Iterating {
+                            visit_last,
+                            next_result,
+                            parents,
+                        };
+
+                        return self.poll_next(_cx);
+                    }
+
+                    let key_from_parents = key_from_nibble_iter(nibble_iter_from_parents(&parents));
+                    let mut visit_last = key == &key_from_parents;
+
+                    if (key.as_slice()) < (key_from_parents.as_slice()) {
```

Comment:
Clippy should have removed these brackets for me. Leaving a comment here for myself